Documentation/design/assetgeneration.md

```diff
@@ -0,0 +1,122 @@
+# Installer asset generation based on dependency resolution
+
+## Goal
+
+Define generation of assets of varied types in installer based on a dependency graph.
+
+## Overview
+
+The installer generates assets based on the dependency [graph](./dependency.md). To generate an asset, all its dependencies have to be generated. The installer needs to define how assets declare this dependency and how it resolves the graph when asked to generated specific targets. Targets are chosen assets that need to be generated, the installer generates all assets that the target depends on and writes to disk only the targets, consuming any inputs from disk or state from a previous run.
+
+Each asset is individually responsible for declaring its dependencies. Each asset is also responsible resolving conflicts when combining its input from disk and its state from a previous run. The installer ensures all the dependencies for an asset is generated and provides the asset with latest state to generate its own output.
+
+## State
+
+State is an internal state of the installer. It stores the contents of all the assets that have been generated.
+
+An example of the State:
+
+```go
+// State is map of assets
+type State struct {
+    Root string // hash for the root asset
+
+    Objects map[string]struct{
+        Parents []string // hashes for our parents
+
+        Objects map[string] {
```

Comment:
Shouldn't the parent include everything but the final targets ? 
I would expect some tree-list structure, besides how is this struct related to the `Assets` struct below?

I was thinking something like:
```go
type State struct {
    Assets []Asset; // Final leaf assets
}
type Asset struct {
    Name string
    Parents []Asset
    Dirty bool
    Hash string
    Data []byte
    ...
}
```

Reply:
The state object is an hashmap containing all the assets, not just the final. Each asset will use this object to fetch its dependencies.